File: js/src/lv1/q2.js
/*
# 크레인 인형뽑기 게임
- 경로: https://school.programmers.co.kr/learn/courses/30/lessons/64061
배치된 인형 정보에서 정해진 순서로 인형을 뽑고, 연속된 인형 종류일 경우 제거한다.
모든 순서를 처리한 후 제거된 인형의 갯수를 리턴한다.

입출력 정보
- 입력(board): 인형 배치도
- 입력(moves): 크레인 움직임
- 결과: 사라진 인형 갯수

제한 사항
- board 배열은 2차원 배열로 크기는 "5 x 5" 이상 "30 x 30" 이하입니다.
- board의 각 칸에는 0 이상 100 이하인 정수가 담겨있습니다.
- 0은 빈 칸을 나타냅니다.
- 1 ~ 100의 각 숫자는 각기 다른 인형의 모양을 의미하며 같은 숫자는 같은 모양의 인형을 나타냅니다.
- moves 배열의 크기는 1 이상 1,000 이하입니다.
- moves 배열 각 원소들의 값은 1 이상이며 board 배열의 가로 크기 이하인 자연수입니다.
*/

export const example = [
    [[[0,0,0,0,0],[0,0,1,0,3],[0,2,5,0,1],[4,2,4,4,2],[3,5,1,3,1]], [1,5,3,5,1,2,1,4], 4]
];

export function solution(board, moves) {
    let removeCount = 0;
    // let emptyCols = [];
    let basket = [];

    moves.forEach(move => {
        let idx = move-1;
        // if (emptyCols.indexOf(idx) > -1) {
        //     return;
        // }

        // let picked = false;
        for (let i = 0; i < board.length; ++i) {
            const row = board[i];
            if(row[idx] === 0) {
                continue;
            }

            if(basket.length > 0 && basket[basket.length-1] === row[idx]) {
                basket.pop();
                removeCount += 2;
            }
            else {
                basket.push(row[idx]);
            }

            row[idx] = 0;
            // picked = true;
            break;
        }

        // if(picked === false)
        // {
        //     emptyCols.push(idx);
        // }
    });

    return removeCount;
}

File: cs/Lv4/Q2.cs
namespace cs.Lv4;

/*
 * # 징검다리
 * https://school.programmers.co.kr/learn/courses/30/lessons/43236
 *
 *
 * 입력
 * - distance: 출발지부터 도착지까지 거리, 1 <= n <= 1000000000
 * - rocks: 바위 위치, 1 <= length <= 50000
 * - n: 제거할 바위 수, 1 <= n <= rocks.length
 * 
 * 출력
 * 바위를 n개 만큼 제거 후 각 바위 사이의 거리의 최소값 중 가장 큰 값 
 */
public class Q2
{
    public record ExampleCase(int Distance, int[] Rocks, int N, int Result);

    public static readonly ExampleCase[] Examples =
    [
        new(25, [2, 14, 11, 21, 17], 2, 4)
    ];

    private class Path
    {
        private readonly List<int> _points;
        private readonly int _removeCount;

        public Path(int distance, int[] rocks, int n)
        {
            _removeCount = n;
            
            Array.Sort(rocks);
            _points = [..rocks, distance]; // C# 12 이상 문법이므로, programmers에서는 사용할 수 없음
            // _points = new List<int>(rocks) { distance };
        }
        
        public bool CanRemove(int min)
        {
            var prev = 0;
            var removed = 0;

            foreach (var p in _points)
            {
                if (p - prev < min) 
                {
                    ++removed;
                }
                else
                {
                    prev = p;
                }
            }

            return removed <= _removeCount;
        }
    }
    
    public static int Solution(int distance, int[] rocks, int n)
    {
        var path = new Path(distance, rocks, n);
        
        var left = 1;
        var right = distance;
        var answer = 0;

        while (left <= right)
        {
            var mid = (left + right) / 2;
            
            if (path.CanRemove(mid))
            {
                answer = mid;
                left = mid + 1;
            }
            else
            {
                right = mid - 1;
            }
        }
        return answer;
    }
}